charmhub_lp_tools/osci_sync.py

```diff
@@ -0,0 +1,247 @@
+    except (TypeError, KeyError) as ex:
+        logger.warning('charm_build_name not found in osci.yaml: %s' % ex)
+        raise CharmNameNotFound()
+    raise CharmNameNotFound()
+
+
+def get_project_vars(osci: OsciYamlType) -> Dict[str, Any]:
+    """Get 'vars' section from osci.yaml.
+
+    :param osci: the osci configuration set in the repo.
+    :returns: vars section
+    """
+    for section in osci:
+        if 'project' not in section:
+            continue
+
+        return section['project']['vars']
+
+
+def gen_auto_build_channel(auto_build_channels: Dict[str, Any],
+                           project_vars: Dict[str, Any],
+                           auto_build_keys: List[Tuple[str, str, object]],
+                           ) -> Dict[str, str]:
+    """Generate a auto_build_channel compatible map.
+
+    Generate a auto_build_channel map based on the information available in
+    ``project_vars`` for the value referenced by ``osci_key``.
+
+    :param auto_build_channels: dictionary with the auto build channels to
+                                compare against.
+    :param project_vars: the project vars declared in osci.yaml
+    :param auto_build_keys: list of tuples where each tuple contains the key
+                            that should be accessed from the
+                            auto_build_channels map, the key that should be
+                            accessed from the project_vars map and finally a
+                            default value if the key is not found in
+                            project_vars.
+    :returns: A new dictionary with the updated values.
+
+    """
+    new_auto_build_channel = deepcopy(auto_build_channels)
+    for lp_key, osci_key, default in auto_build_keys:
+        new_value = project_vars.get(osci_key, default)
+        if auto_build_channels.get(lp_key) != new_value:
+            logger.debug('%s channel from %s to %s',
+                         lp_key,
+                         auto_build_channels.get(lp_key),
+                         new_value)
+            new_auto_build_channel[lp_key] = new_value
+
+    return new_auto_build_channel
+
+
+def setup_parser(subparser: argparse.ArgumentParser):
+    parser = subparser.add_parser(
+        'osci-sync',
+        help='Sync the config defined in osci.yaml to Launchpad.',
+    )
+    parser.add_argument(
+        '--i-really-mean-it',
+        dest='i_really_mean_it',
+        action='store_true',
+        default=False,
+        help=('This flag must be supplied to indicate that the sync/apply '
+              'command really should be used.'),
+    )
+    parser.add_argument(
+        '--repo-dir',
+        dest='repo_dir',
+        default=os.getcwd(),
+        metavar='DIR',
+        help=('Path to the git repository where the charm is located at, by '
+              'default it uses the current working directory.'),
+    )
+    parser.set_defaults(func=main)
+    return parser
+
+
+def main(args: argparse.Namespace,
```

Comment:
I guess that this code came from a separate script, but I'd definitely be inclined to rename the function to something like 'osci_sync' or similar?

Reply:
I'm aiming to have 1 module per subcommand since we have too many subcommands already, maybe we could rename it from 'main' to 'run'?, I'm trying to make it obvious this is an entry point.

Reply:
I'd still be inclined to call it 'osci_sync' as that's what the function does; 'main' isn't really descriptive.  It's actually called via line 186 (tortuously, I agree), so it's not like the function is actually 'accessed' or named from outside of the module.